### DISCUSSION TITLE
Add list of (upcoming) events to the calendar entity base class

### DISCUSSION THREAD

**Discussion by Olen** [home-assistant/architecture#806] [Entity Models] [answered]
## Context

There is currently no way to "loop through" calendar events in e.g a template for an automation or a script.
You can access the current, or next upcoming, event directly in the calendar entity, but that is about it.
There are API endpoints that will list events between dates, but there is no easy way to use these in a template.

Simple use cases are for instance a TTS notification that reads out loud all of todays calendar events in the morning, or an automation with a _condition_ that an event with certain title is happening sometime in the next few days, without necessarily triggering the automation based on the time that event actually starts.

## Proposal

Add a function to the CalendarEntry or the state object so it would be possible to create something like
```
{% for event in calendar.my_calendar.get_events(from=timestamp, to=timestamp) %}
    ...
{% endfor %}
```

**Comment by allenporter** [comment]
There is some discussion about this problem in a previous discussion where we fleshed out some of what it might look like to allow templates for integrations:

https://github.com/home-assistant/architecture/discussions/777#discussioncomment-3042162

I implemented a proof of concept and got it working and I think the biggest part of the change is making all the template stuff, but I think it works.

There wasn't any engagement on the discussion but I think solving this is important to make calendar useful.

**Reply by Olen** [reply]
Thanks.
It looks like we have many of the same or similar ideas. Is there anything I can do to help get some traction on #777?

**Reply by allenporter** [reply]
Have you also considered the implementation details at the next level of detail? IMO having some more confidence that asyncio in templates is not a disaster would help. Right now it looks like a big change and I worry about introducing a lot of complexity there. 

So with that other proposal the API would look something like this:
```
{% for event in  calendar.get_events("calendar.my_calendar", start, end) %}

{% endfor %}
```
I also wonder if perhaps there is a more natural syntax possible so maybe that is something to think through.

**Reply by Olen** [reply]
> IMO having some more confidence that asyncio in templates is not a disaster would help. Right now it looks like a big change and I worry about introducing a lot of complexity there.

I agree that asyncio in templates might have some unwanted consequences, depending on how it is implemented. 
Without being an async-guru, I guess the greatest fear is that a template with custom functions could potentially take a long time to complete rendering, and also that introducing more functions - and especially functions that can be created by any integration - into templating could possibly cause weird errors to creep into anywhere the templating engine is used.
We should probably at least define some "rules", like there are already rules about not polling external sources directly in the `state()`-function of an entity etc., although I am not sure how they can be enforced in custom components.

Another rule should probably be that a function called within a template should never modify the state engine. That could probably lead to some really nasty loops, where the template is set to be re-rendered because the state of one of the entities used in the template is changed from within the template itself. 

So ref #777 - a service call can be used to modify entities states, attributes and config, but a template function must only read (and possibly parse in different ways) data.

I don't know the templating engine well enough to reassure that it will not cause other major issues, but there is a function in the template.py that looks like a starting point to at least ensure that templates should not use too much time to render:

https://github.com/home-assistant/core/blob/e7b594b5cf70ce454d1b3cfa007aa78fa3cfb1b8/homeassistant/helpers/template.py#L464

It seems to be used in the websocket api:
https://github.com/home-assistant/core/blob/a19a7e64d5c179af6444f184d7449a34746aaf19/homeassistant/components/websocket_api/commands.py#L456

So this might be a way to ensure that at least a template is rendered within a reasonable time.

> So with that other proposal the API would look something like this:
> ```
> {% for event in  calendar.get_events("calendar.my_calendar", start, end) %}
> 
> {% endfor %}
> ```
> I also wonder if perhaps there is a more natural syntax possible so maybe that is something to think through.

I don't think that looks too bad.  You could have some sane defaults for start and end, and it would just be a simple `calendar.get_events("calendar.my_calendar")` in most cases, and just let e.g. a check inside the for-loop handle any limits you want to enforce if the default start and end return too many events or events too far into the future for instance.  
It looks slightly odd with the repeated "calendar.", but I think it is better than any variation of e.g. `calendar.get_events("my_calendar")` and having the function add the domain back.

One possibility could be to create wrapper/helper-functions for some common scenarios (these are just wild thoughts, so don't take them too literary
```
calendar.get_next_n_events("calendar.my_calendar", n=10)       # Gets the next 10 events (only future or current)
calendar.get_next_days_events("calendar.my_calendar", days=7) # Gets all the events for today and the upcoming 6 days
calendar.get_this_weeks_events("calendar.my_calendar")         # Gets all events for the currrent week (both future and past)
...
```
I feel it's more intuitive that "days=7" means _"7 days forward including today"_ rather than _"the rest of today and 7 more days"_

Or you could just allow for more optional arguments to the same function and let them be combined
```
calendar.get_events("calendar.my_calendar", max=10)
calendar.get_events("calendar.my_calendar", days=7)
calendar.get_events("calendar.my_calendar", start=dateThisMonday, days=7)
calendar.get_events("calendar.my_calendar", days=7, max=10)
...
```

```
def get_events(entity_id, start=None, end=None, days=7, max=10): 
  if start is None: 
    start = datetime.now().replace(hour=0, minute=0, second=0, microsecond=0)

  if end is None:
    end = start + timedelta(days=days)    

  return await async_get_events(start, end)[:max]
```

That way people don't need to calculate "today(@midnight) + 7 days" in each and every template.

**Comment by allenporter** [comment, marked as answer]
We now have a solution for https://github.com/home-assistant/architecture/discussions/777#discussioncomment-6127898 and have implemented the first version of services with return values to solve this and we'll have a way to return events in https://github.com/home-assistant/core/pull/94759 -- as a result, i'll close this discussion.
